### PR TITLE
Use defaults when decoding JSON metadata

### PR DESCRIPTION
**Pull request by benjeffery**
A start on #1073 

So far this only does defaults on objects, and objects in objects, but does not do arrays. For example an array of objects could have a default key for all the instances in the array, or an array itself have a default value.

Note that where a deeper default exists it overrides a shallower one. This also means that if a shallower default of an object has many keys, and one of those keys has a deeper default, all of the shallow default will be ignored.

# PR Checklist:
- [x] Tests that fully cover new/changed functionality.
- [ ] Documentation including tutorial content if appropriate.
- [ ] Changelogs, if there are API changes.

### CODE REVIEW

File: python/tests/test_metadata.py [hunk 1]
```diff
@@ -515,6 +515,68 @@ def decode(self, encoded):
...
+        schema = {
+            "codec": "json",
+            "type": "object",
+            "properties": {
+                "obj": {
+                    "type": "object",
+                    "properties": {
+                        "number": {"type": "number", "default": 5},
+                    },
+                    "default": {"number": 42},  # This will be ignored
```
Comment (jeromekelleher) [inline python/tests/test_metadata.py:572]: Doesn't this default violate the schema anyway? 42 isn't an object, is it?

File: python/tskit/metadata.py [hunk 2]
```diff
@@ -109,13 +110,33 @@ def register_metadata_codec(
 
 class JSONCodec(AbstractMetadataCodec):
     def __init__(self, schema: Mapping[str, Any]) -> None:
-        pass
+        # Recurse through the schema finding default values to fill in on decode
+        self.defaults = {}
+
+        def record_defaults(path, sub_schema):
```
Comment (jeromekelleher) [inline python/tskit/metadata.py:116]: Elegant!

File: python/tskit/metadata.py [hunk 3]
```diff
@@ -109,13 +110,33 @@ def register_metadata_codec(
...
+
+        record_defaults((), schema)
 
     def encode(self, obj: Any) -> bytes:
         return tskit.canonical_json(obj).encode()
 
     def decode(self, encoded: bytes) -> Any:
-        return json.loads(encoded.decode())
+        result = json.loads(encoded.decode())
+        # Assign default values
```
Comment (jeromekelleher) [inline python/tskit/metadata.py:130]: I think it's logically simpler to start with the ``defaults`` dictionary, and write in anything that's in the decoded version, so
```python
result = copy.deepcopy(self.defaults)
recursive_update(result, decoded)
```

I agree this is made substantially trickier by the nestedness requirements.

I'm tempted to just implement this for top-level values initially (throwing an error if someone tries to set a default for a nested type), and see how far it gets us. It'll be a lot more work up front if we decide to support all possible combinations of nested defaults.

**Review by jeromekelleher — COMMENTED** [review]
Looks good, and very elegant code. In the interest of not getting bogged down on details, though, I wonder if we should do something more limited in the first instance, though.

**Comment by jeromekelleher** [comment]
Bumping my vote for just doing this on top-level stuff and raising an error if someone tries to do anything more complicated with defaults...

**Comment by benjeffery** [comment]
Closed in favour of #1272

**Pull request closed without merging**